elb_{network_lb,target_group}.py: allowing UDP and TCP_UDP protocols

mulatinho opened the pull request:
Allow UDP, TLS, TCP_UDP protocols as described on AWS SDK.

- Fixing documentation
- Making health checks with response codes and paths only in HTTP/HTTPS
- https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/elbv2.html
    
Fixes: #65265

mulatinho committed:
elb_{network_lb,target_group}: allow UDP/TCP_UDP

elb_network_lb.py: allow UDP and TCP_UDP protocols

- Fixing documentation
- Add support to UDP and TCP_UDP as described on AWS SDK

elb_target_group.py: allow UDP, TLS, TCP_UDP proto

- Fixing documentation
- Making health checks with response codes and paths only in HTTP/HTTPS
- Allow UDP, TLS, TCP_UDP protocols as described on AWS SDK.

others:
- Added changelog fragments
- Integration test

Fixes: #65265
Signed-off-by: Alexandre Mulatinho <alex@mulatinho.net>

mulatinho commented:
ready_for_review

Akasurde commented:
@jillr @s-hertel Could you please review this? Thanks.

jillr commented:
Could you please add integration tests for the new protocols, and a changelog? Thanks!

mulatinho commented:
hey @jillr thanks for the first review, I added the changelog fragment and also added tests into network load balancer and target groups using UDP protocols. please, review when you can. :)

mulatinho commented:
ready_for_review

mulatinho commented:
ready_for_review

jillr commented on test/integration/targets/elb_target/playbooks/roles/elb_target/tasks/main.yml:
```diff
@@ -213,6 +226,19 @@
 
+    - name: remove an udp unused target
+      elb_target:
+        target_group_name: "{{ tg_name }}-udp"
```
There are no targets in this group, and `target_id` is a required argument to `elb_target`.  I think you mean to use `elb_target_group` to delete the UDP target group here?

mulatinho replied:
thank you for point me to right path! :D

jillr commented on test/integration/targets/elb_network_lb/tasks/test_creating_nlb.yml:
```diff
@@ -27,13 +27,23 @@
+          Port: 17
+          DefaultActions:
+            - Type: forward
+              TargetGroupName: "{{ tg_name }}"
```
The target group created in `test/integration/targets/elb_network_lb/tasks/main.yml` is TCP-only, and the TLS listener in this task can not go to a TCP_UDP target group.  You'll need to create a second TCP_UDP target group in `test/integration/targets/elb_network_lb/tasks/main.yml` and a second NLB to test the new protocols, then be sure all the additional resources are cleaned up in their respective task file's `always` block.

mulatinho replied:
thank you for point me to right path! :D

jillr reviewed (changes requested):
Thanks @mulatinho! Tests are failing for me though, a couple things that should help.

mulatinho commented:
hey @jillr thank you for point me to right path! :D 
can you please review it again ? I think everything is fine now :)

jillr commented on test/integration/targets/elb_target/playbooks/roles/elb_target/tasks/main.yml:
```diff
@@ -102,6 +102,19 @@
 
+    - name: set up testing target group (type=instance) with UDP protocol
+      elb_target_group:
+        name: "{{ tg_tcpudp_name }}"
```
this variable will need to be created for this test suite in test/integration/targets/elb_target/playbooks/roles/elb_target/defaults/main.yml

mulatinho replied:
oh man, sorry :D

jillr commented on test/integration/targets/elb_target/playbooks/roles/elb_target/tasks/main.yml:
```diff
@@ -213,6 +226,20 @@
+    - name: target group was deleted
+      assert:
+        that:
+          - result.changed
```
This assertion is failing because there are no targets to be removed from `tg_tcpudp_name`.  Are you trying to delete the target group itself?  That would need to be done with `elb_target_group`.

mulatinho replied:
hmmm, fixed that! I think its good now

jillr reviewed (changes requested):
Thanks @mulatinho, couple things from testing this latest patchset.

mulatinho commented:
ready_for_review

jillr reviewed (approved):
LGTM and tests are passing - thanks @mulatinho!

threewordphrase commented:
@jillr Do you know what release this will be in?
